DragaliaAPI.Test/Features/Shop/PaymentServiceTest.cs

```diff
@@ -0,0 +1,170 @@
+﻿using DragaliaAPI.Database.Entities;
+using DragaliaAPI.Database.Repositories;
+using DragaliaAPI.Features.Shop;
+using DragaliaAPI.Models;
+using DragaliaAPI.Models.Generated;
+using DragaliaAPI.Services.Exceptions;
+using DragaliaAPI.Shared.Definitions.Enums;
+using DragaliaAPI.Test.Utils;
+using MockQueryable.Moq;
+
+namespace DragaliaAPI.Test.Features.Shop;
+
+public class PaymentServiceTest
+{
+    private readonly PaymentService paymentService;
+    private readonly Mock<IUserDataRepository> mockUserDataRepository;
+
+    public PaymentServiceTest()
+    {
+        this.mockUserDataRepository = new(MockBehavior.Strict);
+        this.paymentService = new(
+            LoggerTestUtils.Create<PaymentService>(),
+            this.mockUserDataRepository.Object
+        );
+    }
+
+    [Theory]
+    [InlineData(PaymentTypes.Wyrmite, 500, 500)]
+    [InlineData(PaymentTypes.HalidomHustleHammer, 1, 10)]
+    [InlineData(PaymentTypes.Coin, 100000, 10000000)]
+    [InlineData(PaymentTypes.ManaPoint, 52013, 52014)]
+    public async Task ProcessPayment_ValidTarget_SubtractsCurrency(
+        PaymentTypes type,
+        int cost,
+        int total
+    )
+    {
+        this.mockUserDataRepository
+            .SetupGet(x => x.UserData)
+            .Returns(
+                new List<DbPlayerUserData>()
+                {
+                    new()
+                    {
+                        DeviceAccountId = IdentityTestUtils.DeviceAccountId,
+                        BuildTimePoint = type == PaymentTypes.HalidomHustleHammer ? total : 0,
+                        Coin = type == PaymentTypes.Coin ? total : 0,
+                        Crystal = type == PaymentTypes.Wyrmite ? total : 0,
+                        ManaPoint = type == PaymentTypes.ManaPoint ? total : 0,
+                    }
+                }
+                    .AsQueryable()
+                    .BuildMock()
+            );
+
+        await this.paymentService.ProcessPayment(type, new PaymentTarget(total, cost));
+
+        this.mockUserDataRepository.VerifyAll();
```

Comment:
If you assigned the UserData here I think it would get modified by reference and you could check the subtraction